### ISSUE TITLE
Running Pester test from within a module fails to test a private function.

### ISSUE THREAD

**Issue by Marc013** [pester/Pester#2003]
Can you please tell me what I am doing wrong?

I’m creating my own pwsh module with some functions I find handy.
One of those functions is for generating unit test and code coverage HTML reports (New-UnitTestReport).

While developing function New-UnitTestReport I am testing it using the Pester tests I’ve created for module LazyGuy.
Module LazyGuy requires private function Invoke-Butler, the module does not function without it. Hence, I have created helper function Assert-HelperFunction to validate if private function Invoke-Butler is present when any of the public functions of the module is called. Each public functions calls Assert-HelperFunction.

The fictitious module:

```TXT
C:\GIT\FANCYSTUFF\MODULES
└─── LazyGuy
    │   LazyGuy.psd1
    │   LazyGuy.psm1
    │
    ├───private
    │       Assert-HelperFunction.ps1
    │       Invoke-Butler.ps1
    │
    └───public
            Clear-MyDesk.ps1
            Get-Coffee.ps1
            New-Beer.ps1
```

Assert-HelperFunction:

```PowerShell
function Assert-HelperFunction {
    <#
    .SYNOPSIS
        Throw error if a helper function is not available.

    .DESCRIPTION
        Assert-HelperFunction asserts if the helper function is available during execution.
    #>
    [string]$ModuleName = 'LazyGuy'
    [string[]]$FunctionNames = 'Invoke-Butler'
    $Module = Get-Module -Name $ModuleName

    if ([string]::IsNullOrWhiteSpace($Module)) {
        throw "Missing module '$ModuleName'"
    }

    $ModuleFunctions = $Module.Invoke( { Get-Command -Module $ModuleName } )

    foreach ($FunctionName in $FunctionNames) {
        $Dependency = $ModuleFunctions | Where-Object -Property Name -EQ $FunctionName

        if ([string]::IsNullOrWhiteSpace($Dependency)) {
            [string[]]$MissingFunction += $FunctionName
        }
    }
    if (-not [string]::IsNullOrWhiteSpace($MissingFunction)) {
        throw "Missing dependent function '$($MissingFunction -join ' & ')'"
    }
}
```

Function Assert-HelperFunction has its own Pester test.

Assert-HelperFunction.tests.ps1:
```PowerShell
$ModuleName = 'LazyGuy'
$ModulePath = Join-Path -Path $PSScriptRoot -ChildPath "..\modules\$ModuleName"
Remove-Module $ModuleName -ErrorAction SilentlyContinue
Import-Module -Name $ModulePath

Describe 'Function Assert-HelperFunction' {
    Context 'asserting function' {
        It 'should call throw default error Execution failed' {
            InModuleScope 'LazyGuy' {
                Set-Variable -Name MissingFunction -Value mock -Scope Script

                { Assert-HelperFunction } | Should -Throw "Missing dependent function 'mock'"
            }
        }
    }
}
```
This test runs successfully when running command ‘Invoke-Pester -Path C:\git\FancyStyff\tests\Assert-HelperFunction.tests.ps1’

However, when I run this test via module function New-UnitTestReport I get error:

> Running tests from 'C:\git\FancyStuff\tests\Assert-HelperFunction.tests.ps1'
> Describing Function Assert-HelperFunction
>  Context asserting function
>    [-] should call throw default error Execution failed 26ms (19ms|7ms)
>     Expected an exception, with message 'Missing dependent function 'mock'' to be thrown, but the message was 'Missing module 'LazyGuy''. from C:\git\LazyGuy\modules\FancyStuff\private\Assert-HelperFunction.ps1:14 char:9
>         +         throw "Missing module '$ModuleName'"
>         +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     at { Assert-HelperFunction } | Should -Throw "Missing dependent function 'mock'", C:\git\FancyStuff\tests\Assert-HelperFunction.tests.ps1:12
>     at <ScriptBlock>, C:\git\FancyStuff\tests\Assert-HelperFunction.tests.ps1:12

When I convert the function to a script this error is not thrown!

Module function New-UnitTestReport (it’s work in progress):

```PowerShell
function New-UnitTestReport {
    <#
    .SYNOPSIS

    .DESCRIPTION

    .EXAMPLE
    $Param = @{
        Path                = 'C:\git\FancyStuff\tests'\
        ScriptPath          = 'C:\git\FancyStuff\modules\LazyGuy\private\*.ps1', 'C:\git\FancyStuff\modules\LazyGuy\public\*.ps1'
        ReportUnitPath      = 'C:\ReportUnit\ReportUnit.exe'
        ReportGeneratorPath = 'C:\ReportGenerator\ReportGenerator.exe'
        ReportType          = 'HtmlInline_AzurePipelines', 'HtmlInline_AzurePipelines_Dark'
        ReportTitle         = 'Nice Stuff!!'
        ShowReport          = $true
    }
    New-UnitTestReport @Param

    This command will . . .
    #>
    [CmdletBinding()]
    param (
        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Directories to be searched for tests, paths directly to test files, or combination of both.')]
        [string]$Path,
        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Directorie where the scripts are located for code coverage.')]
        [array]$ScriptPath,
        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Full path of ReportUnit.exe. (e.g. "C:\ReportUnit\ReportUnit.exe")')]
        [string]$ReportUnitPath,
        [Parameter(
            Mandatory = $true,
            HelpMessage = 'Full path of ReportGenerator.exe. (e.g. "C:\ReportGenerator\ReportGenerator.exe")')]
        [string]$ReportGeneratorPath,
        [Parameter(
            Mandatory = $false,
            HelpMessage = 'Code coverage output format and scope.')]
        [ValidateSet('Badges', 'Clover', 'Cobertura', 'CsvSummary', 'Html', 'HtmlChart',
            'HtmlInline', 'HtmlInline_AzurePipelines', 'HtmlInline_AzurePipelines_Dark',
            'HtmlSummary', 'JsonSummary', 'Latex', 'LatexSummary', 'lcov', 'MHtml',
            'PngChart', 'SonarQube', 'TeamCitySummary', 'TextSummary', 'Xml', 'XmlSummary')]
        [System.Collections.ArrayList]$ReportType = 'Html',
        [Parameter(
            Mandatory = $false,
            HelpMessage = 'Title of the code coverage report.')]
        [string]$ReportTitle = 'Code Coverage',
        [Parameter(
            Mandatory = $false,
            HelpMessage = 'Open HTML report.')]
        [switch]$ShowReport
    )
    #Requires -Modules @{ ModuleName='Pester'; ModuleVersion='5.2.2' }

    $ErrorPrefernce = $ErrorActionPreference
    $ErrorActionPreference = 'SilentlyContinue'
    $AssertReportGenerator = & $ReportGeneratorPath /?
    $ErrorActionPreference = $ErrorPrefernce

    if ([string]::IsNullOrWhiteSpace($AssertReportGenerator)) {
        throw "`nUnable to find ReportGenerator.exe. `nPlease make sure ReportGenerator.exe is installed. `nhttps://danielpalme.github.io/ReportGenerator/"
    }

    # Run Pester test

    ### ToDo: Specify report output directory path. Create path if not exist including dir for test results and code coverage
    ### E.g. c:\testResults\unitTest\ & c:\testResults\codeCoverage\

    $CodeCoverageOutputPath = "$Path/coverage.xml"
    $TestResultOutputPath = "$Path/testResults.xml"

    $configuration = New-PesterConfiguration
    $configuration.Run.Path = $Path
    $configuration.Run.Exit = $true
    $configuration.Should.ErrorAction = 'Continue'
    $configuration.CodeCoverage.Enabled = $true
    $configuration.CodeCoverage.OutputFormat = 'JaCoCo'
    $configuration.CodeCoverage.OutputPath = $CodeCoverageOutputPath
    $configuration.CodeCoverage.Path = $ScriptPath
    $configuration.TestResult.Enabled = $true
    $configuration.TestResult.OutputFormat = 'NUnitXml'
    $configuration.TestResult.OutputPath = $TestResultOutputPath
    $configuration.Output.Verbosity = 'Detailed'

    try {
        Invoke-Pester -Configuration $configuration

        # Generating unit tests report
        & $ReportUnitPath $TestResultOutputPath | Out-String | Out-Null ### ToDo: Add report destinatino path

        # Generating code coverage report
        if ($ReportType.Contains('HtmlInline_AzurePipelines') -and $ReportType.Contains('HtmlInline_AzurePipelines_Dark')) {
            Write-Warning "You specified report type 'HtmlInline_AzurePipelines' and 'HtmlInline_AzurePipelines_Dark'. `nOnly one of these report types can be created at one time. `nRemoving report type 'HtmlInline_AzurePipelines_Dark'`n"
            $ReportType.Remove('HtmlInline_AzurePipelines_Dark')
        }

        [string]$CoverageReportDir = 'coveragereport'
        [string]$ReportType = Join-String -InputObject $ReportType -Separator ';'
        [string]$SourceDirs = $ScriptPath.TrimEnd('*.ps1') | Join-String -Separator ';'

        $CodeCoverageResult = & $ReportGeneratorPath -reports:$CodeCoverageOutputPath -targetdir:$CoverageReportDir -sourcedirs:$SourceDirs -reporttypes:$ReportType -title:$ReportTitle | Out-String
        # $Result

        if ($ShowReport.IsPresent) {
            $DirSeparator = [IO.Path]::DirectorySeparatorChar

            #&

            foreach ($Entry in $CodeCoverageResult.Split(':')) {
                if ($Entry -match 'Writing report file') {
                    [string]$Report = $Entry.Split("'")[1].Split("$DirSeparator")[-1]
                    Write-Host "Operning Report '$CoverageReportDir$DirSeparator$Report'" -ForegroundColor Magenta

                    & $CoverageReportDir$DirSeparator$Report
                }
            }
        }
    }
    finally {
        Remove-Item -LiteralPath $CodeCoverageOutputPath -Force -ErrorAction SilentlyContinue
        Remove-Item -LiteralPath $TestResultOutputPath -Force -ErrorAction SilentlyContinue
    }
}
```

**Comment by nohwnd** [comment]
It looks like `$Module = Get-Module -Name $ModuleName` won't get the module for some reason. Do you have this in a runnable state somewhere? I don't see anything obvious. You should be able to get the module even from the module itself without any problem. So I don't see why it would not resolve, unless it is not imported, which it should be.

**Comment by Marc013** [comment]
Hi @nohwnd ,

Thank you for your feedback.

The Pester test itself runs without any problem when I run it directly via `Invoke-Pester`. However, when it is run via function `New-UnitTestReport` (imported via its own module) it fails.

I've added a `Write-Host "$(Get-Module | Out-String)"` to Pester test `Assert-HelperFunction.tests.ps1` to show which modules are loaded.
Running `Invoke-Pester` the module of function `Assert-HelperFunction` is shown.
When running function `New-UnitTestReport` it isn't shown.
_All other Pester tests run fine._

The strangest thing is that when I copy and paste the function code (`New-UnitTestReport`) directly in the terminal it works like a charm. Only when importing it via the module it fails.

I have the feeling it is a scoping issue. Unfortunately, I have no clue how to fix it. It's quite frustrating.

The code of function `New-UnitTestReport` is available in repo [https://github.com/Marc013/Marc013Stuff](https://github.com/Marc013/Marc013Stuff)

**Comment by nohwnd** [comment]
Yeah, definitely an issue with session states. Could you make a branch in that repo, where I can just run a single script to see this fail? And tell me what exactly I should run. I want to help you, but I can't spend 2 hours trying to figure out how to reproduce this :)

**Comment by Marc013** [comment]
I have create the branch you asked.
> [https://github.com/Marc013/Marc013Stuff/tree/troubleshooting](https://github.com/Marc013/Marc013Stuff/tree/troubleshooting)

In this branch I have commented code that has no relation to the issue preventing you from installing some executables.

I have also created a dummy module that produces the issue reported. 
> [https://github.com/Marc013/FancyStuff](https://github.com/Marc013/FancyStuff)

To run function `New-UnitTestReport` (module `Marc013Stuff`) I have added a description to the README in that repo. It reads:

> Running function _New-UnitTestReport_:
> 
> * Download/clone repo [FancyStuff][1].
>     * In the example below the repo is cloned in path _C:\git_.
> * Download/clone this repo.
>     * In my situation the repo is cloned in path _C:\git_
> * Import module `Marc013Stuff`
>     * `Import-Module -Name C:\git\Marc013Stuff\Marc013Stuff`
> * Run command:
> ```PowerShell
> New-UnitTestReport -Path C:\git\FancyStuff\tests -ScriptPath C:\git\FancyStuff\modules\LazyGuy\public\*.ps1 -ReportUnitPath MockReportUnit.exe -ReportGeneratorPat MockGenerator.exe -ReportType HtmlInline_AzurePipelines, HtmlInline_AzurePipelines_Dark
> ```

[1]: https://github.com/Marc013/FancyStuff

**Comment by nohwnd** [comment]
```powershell
get-module m,p,l | remove-module

new-module -name m -scriptBlock { 
    New-Module -Name p -ScriptBlock { 
        function inModule ($name, $sb) { 
            $m = get-module $name -All
            & ($m) $sb
        }

        function invoke ($sb) { 
            & $sb
        }
    } | Import-Module



    function nutr {
        # this runs in your marco module
        invoke {
            New-Module l -ScriptBlock { 
                function f () { 
                    $l = Get-Module l
                    if (-not $l) { 
                        throw "aaaargh!"
                    }
                }
            } | Import-Module

            inModule "l" { 
                f 
            }
        }
    }
} | Import-Module 

nutr
```

So this is roughly what is happening in your module. You call invoke-pester from the Marco (m) module, your module imports Pester (p), and that imports LazyGuy (l) into the context of Marco (m). Importing a module into a module works just fine, but it seems like there is bug in Get-Module when it is used in the inner modules. It shows modules that are imported into script scope rather than into the parent module. So it does not see itself unless you specify -All. 

```powershell
Get-Module a,b, MMM, GGG | Remove-Module 
new-module -Name MMM -ScriptBlock {} | Import-Module
New-Module -name a -scriptBlock {
    new-module -Name GGG -ScriptBlock {} | Import-Module
    New-Module -Name b -ScriptBlock {
        function Get-B {
            "Modules in B are (incorrectly a, MMM which are modules imported to script scope):"
            Get-Module
        }
    } | Import-Module -Force

    function Invoke-B {
        Get-B
    }

    Write-Host "Modules in A are (b, MMM and GGG):"
    Write-Host (Get-Module | Out-String )
} | Import-Module -Force

Invoke-B 
```

```
Modules in A are (b, MMM and GGG):

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     0.0        b                                   Get-B
Script     0.0        GGG
Script     0.0        MMM

Modules in B are (incorrectly a, MMM which are modules imported to script scope):

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     0.0        a                                   {Get-B, Invoke-B}
Script     0.0        MMM
```

So one solution to your problem is to call Get-Module -All, but imho a better solution is to check if your current module is what you are looking for and use that if it is.

```powershell
$Module = if ($null -ne $ExecutionContext.SessionState.Module -and $ModuleName -eq $ExecutionContext.SessionState.Module.Name) {
    $ExecutionContext.SessionState.Module
}
else { 
    Get-Module $ModuleName -All
}
```

**Comment by nohwnd** [comment]
^^^ @fflaten this a piece of module trivia that might come handy to know some time later.

**Comment by Marc013** [comment]
@nohwnd 
Thank you very much for this elaborate explanation!
It really helps to understand the what is happening instead of only getting the solution 🤘🏽

**Comment by nohwnd** [comment]
You are welcome, I was recently solving an issue that had the same symptoms in Should -Invoke, or at least I thought it did. Originally I could not see the module in module, but then later I could see the module in the Get-Module list. Until now I did not know how I got into that first state. And I was not able to reproduce it later. Now I still don't know how I got into that first state, but now I know what caused it. 

So we both learned here, as it is usually the case when a mysterious issue with scopes is solved 😁
